### ISSUE TITLE
False positive on Network Policies .. Sometimes

### ISSUE THREAD

**Issue by bpfoster** [derailed/popeye#311]
<img src="https://raw.githubusercontent.com/derailed/popeye/master/assets/popeye_bug.png" align="right" width="100" height="auto"/>

<br/>
<br/>
<br/>

**Describe the bug**
Running popeye sometimes reports
```
  · foo/foo-75f84cd95b-9s62d...................................................😱
    😱 [POP-1204] Pod Ingress is not secured by a network policy.
    😱 [POP-1204] Pod Egress is not secured by a network policy.
```
Despite there being a Network Policy for it.

**Expected behavior**
No warning on missing Network Policy

**Versions (please complete the following information):**
 - OS: RHEL 8
 - Popeye: v0.21.3
 - K8s: v1.26.11

**Additional context**
I have created NetworkPolicies using the 
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
```
resource type.

I notice the cluster also has Calico installed providing their NetworkPolicy CRD:
```
$ kubectl get crd | grep networkpolicies
globalnetworkpolicies.crd.projectcalico.org
networkpolicies.crd.projectcalico.org
```

Installed policies:
```
$ kubectl get networkpolicies.crd.projectcalico.org 
No resources found in foo namespace.

$ kubectl get networkpolicies.networking.k8s.io 
NAME                                     POD-SELECTOR
foo                 app.kubernetes.io/instance=foo,app.kubernetes.io/name=foo
```

When I debug popeye, I see that the GVR being queried in the DB for network policies is sometimes the calico one, and sometimes the k8s.io one.
https://github.com/derailed/popeye/blob/d09ec25f3834d2c6a171486b9726b0a91793e3f0/internal/lint/pod.go#L96

When it's the Calico one, the NetworkPolicy is not found.

My guess is there's some non-determinism in the order in which the CRDs are returned/loaded, and from what I can see, the popeye DB only supports 1 per type and is overwriting whichever comes first with the second one.
https://github.com/derailed/popeye/blob/d09ec25f3834d2c6a171486b9726b0a91793e3f0/internal/alias.go#L92-L116

Thinking about it... should popeye know the GVR for each resource type it cares about, and ignore others?  In this case it knows it wants `networkpolicies.networking.k8s.io` and would discard the caclico type.  If there was a calico NP (or some other random one), popeye would panic when it tries to assert to `netv1.NetworkPolicy`.

**Comment by bpfoster** [comment]
This is maybe out of scope for this issue, and admittedly I'm still working on understanding Network Policies, but I'm struggling to understand the logic in Popeye on checking ingress/egress policies.

If we get to this branch of code:
https://github.com/derailed/popeye/blob/d09ec25f3834d2c6a171486b9726b0a91793e3f0/internal/lint/pod.go#L119-L126

I'm not sure the logic in `checkIngresses` / `checkEgresses` is wholly correct.

https://github.com/derailed/popeye/blob/d09ec25f3834d2c6a171486b9726b0a91793e3f0/internal/lint/pod.go#L177-L199

If we have a NetworkPolicy which targets the pod in question, and it has ingress rules, `checkIngresses` appears to want the ingress's `from` to point to the same pod.

Similarly for `checkEgresses`, it's wanting the egress's `to` to point to the same pod.

In other words, this seems to only pass when the NetworkPolicy says the pod can talk to itself.

**Comment by derailed** [comment]
@bpfoster For the initial issue, this is indeed a problem. Will fix on the next drop.
For the later, this code checks if a given pod is protected by a networkpolicy.

**Comment by derailed** [comment]
Fixes v0.21.5

**Comment by siegy22** [comment]
I just had a similar thing come up when denying egress:

```
  · about-me/about-me-5d8f856785-pb7vx.............................................................😱
    😱 [POP-1204] Pod Egress is not secured by a network policy.
```

This is my NetworkPolicy

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: deny-egress
spec:
  podSelector:
    matchLabels:
      app: about-me
  policyTypes:
    - Ingress
    - Egress
  ingress:
    - {}
  egress: []
```

I did verify that the policy works by looking at the drops in Cilium. 

Am I missing something?

**Comment by bpfoster** [comment]
I hadn't revisited this.  I too am still getting false positives on networkpolicy linting (on latest popeye v0.21.5).

In my case, I get `Pod Ingress is not secured by a network policy` despite NetworkPolicies being present.  `checkIngresses()` is still coming back with no matches so I still wonder about how that is being evaluated.
My Network Policies look like:
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  labels:
    app.kubernetes.io/instance: my-instance
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: my-service
    app.kubernetes.io/version: 0.0.1
    helm.sh/chart: my-service-0.1.0
  name: my-networkpolicy
  namespace: my-namespace
spec:
  egress:
  - ports:
    - port: 53
      protocol: UDP
    - port: 53
      protocol: TCP
  - to:
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: my-namespace
  - ports:
    - port: 80
      protocol: TCP
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: my-namespace
    ports:
    - port: http
      protocol: TCP
  podSelector:
    matchLabels:
      app.kubernetes.io/instance: my-instance
      app.kubernetes.io/name: my-service
  policyTypes:
  - Ingress
  - Egress
```


For @siegy22 's example, playing around with some unit tests, changing `egress` to look like `ingress` removes the finding:
```
ingress:
  - {}
egress:
  - {}
```
But I frankly don't know if the result is the same?

**Comment by siegy22** [comment]
@bpfoster Your example denies ingress **and** egress. My goal is to have any ingress and no egress.

Which is defined by:

```
ingress:
  - {}
egress: []
 ```
 
But popeye says there's no egress defined (which in theory is correct) but the effect is not the same as not defining it.

**Comment by bpfoster** [comment]
@derailed please see PR #367 - I think this will address the false positives @siegy22 and I are seeing, if I have not misunderstood something here.
